File: 14/day14.py
#!/usr/bin/env python3

import re


def main():
    with open("input", "r") as f:
        lines = f.read().strip().splitlines()

    instructions = parse(lines)
    print(solve(instructions, version=1))
    print(solve(instructions, version=2))


def parse(lines):
    instructions = []
    for l in lines:
        if l.startswith("mask"):
            instructions.append(l.split(" = ", 2)[-1])
        elif l.startswith("mem"):
            instructions.append(tuple(map(int, re.findall(r"\d+", l))))
    return instructions


def solve(instructions, version=1):
    mem = {}
    mask = ("X" if version == 1 else "0") * 36

    for i in instructions:
        if "X" in i:
            mask = i
        else:
            if version == 1:
                mem[i[0]-1] = apply_mask(mask, i[1])
            elif version == 2:
                for j in calculate_floating(mask, i[0]):
                    mem[j-1] = i[1]

    return sum(mem.values())


def apply_mask(mask, val):
    result = 0
    for i, j in zip(mask, list(bits(val, pad=len(mask)))[::-1]):
        result = (result << 1) ^ (j if i == "X" else int(i))
    return result


def calculate_floating(mask, mem):
    mem_bits = list(bits(mem, pad=len(mask)))[::-1]
    if mask[0] == "X":
        results = [0, 1]
    else:
        m = int(mask[0])
        results = [m if m == 1 else mem_bits[0]]

    for i, j in zip(mask[1:], mem_bits[1:]):
        if i == "X":
            for r in range(len(results)):
                v = results[r]
                results[r] = (v << 1) ^ 0
                results.append((v << 1) ^ 1)
        else:
            m = int(i)
            a = m if m == 1 else j
            for r in range(len(results)):
                v = results[r]
                results[r] = (v << 1) ^ a

    return results


def bits(n, pad=None):
    if pad is None:
        pad = n.bit_length()

    count = 1
    while n or count <= pad:
        yield n & 1
        n >>= 1
        count += 1


if __name__ == "__main__":
    main()

File: 14/day14.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type action int

const (
	actionMask  action = 0
	actionWrite action = 1
	actionNop   action = 2
)

type instruction struct {
	action  action
	mask    string
	address int
	value   int
}

func main() {
	instructions, err := parse("input")

	if err != nil {
		os.Exit(1)
	}

	fmt.Println(solve(instructions, 1))
	fmt.Println(solve(instructions, 2))
}

func solve(instructions []instruction, version int) int {
	mem := map[int]int{}
	mask := ""

	for _, i := range instructions {
		switch i.action {
		case actionMask:
			mask = i.mask

		case actionWrite:
			switch version {
			case 1:
				mem[i.address] = applymask(mask, i.value)

			case 2:
				for _, v := range calculateFloats(mask, i.address) {
					mem[v] = i.value
				}
			}
		}
	}

	sum := 0
	for _, v := range mem {
		sum += v
	}

	return sum
}

func applymask(mask string, value int) int {
	if len(mask) == 0 {
		return value
	}

	n := applymask(mask[:len(mask)-1], value>>1)

	if mask[len(mask)-1] == 'X' {
		n <<= 1
		n ^= value & 1
	} else {
		n <<= 1
		n ^= atoi(mask[len(mask)-1:])
	}

	return n
}

func calculateFloats(mask string, address int) []int {
	if len(mask) == 1 {
		if mask[0] == 'X' {
			return []int{0, 1}
		}

		m := atoi(mask[len(mask)-1:])
		if m == 0 {
			m = address & 1
		}
		return []int{m}
	}

	a := calculateFloats(mask[:len(mask)-1], address>>1)

	if mask[len(mask)-1] == 'X' {
		aLen := len(a)
		for i := 0; i < aLen; i++ {
			v := a[i] << 1
			a[i] = v ^ 0
			a = append(a, v^1)
		}
	} else {
		m := atoi(mask[len(mask)-1:])
		if m == 0 {
			m = address & 1
		}

		for i, v := range a {
			a[i] = (v << 1) ^ m
		}
	}

	return a
}

func parse(path string) ([]instruction, error) {
	instructions := []instruction{}
	file, err := os.Open(path)

	if err != nil {
		return instructions, err
	}

	defer file.Close()

	re := regexp.MustCompile(`\d+`)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()

		action := actionNop
		mask := ""
		address := 0
		value := 0

		if strings.HasPrefix(line, "mask") {
			action = actionMask
			mask = line[len(line)-36:]
		} else {
			action = actionWrite
			vals := re.FindAllString(line, 2)
			address = atoi(vals[0])
			value = atoi(vals[1])
		}

		instructions = append(instructions, instruction{
			action:  action,
			mask:    mask,
			address: address,
			value:   value,
		})
	}

	return instructions, scanner.Err()
}

func atoi(s string) int {
	n, err := strconv.Atoi(s)
	if err != nil {
		return -1
	}
	return n
}
